addons/stock/models/stock_production_lot.py

```diff
@@ -192,27 +192,43 @@ def action_lot_open_transfers(self):
     def _find_delivery_ids_by_lot(self, lot_path=None, delivery_by_lot=None):
         if lot_path is None:
             lot_path = set()
-
         domain = [
             ('lot_id', 'in', self.ids),
             ('state', '=', 'done'),
             '|', ('picking_code', '=', 'outgoing'), ('produce_line_ids', '!=', False)
         ]
         move_lines = self.env['stock.move.line'].search(domain)
+        moves_by_lot = {
+            lot_id: {'producing_lines': set(), 'barren_lines': set()}
+            for lot_id in move_lines.lot_id.ids
+        }
+        for line in move_lines:
+            if line.produce_line_ids:
+                moves_by_lot[line.lot_id.id]['producing_lines'].add(line.id)
+            else:
+                moves_by_lot[line.lot_id.id]['barren_lines'].add(line.id)
         if delivery_by_lot is None:
             delivery_by_lot = dict()
         for lot in self:
             delivery_ids = set()
-            if lot.id in lot_path:
-                continue
-            for line in move_lines.filtered(lambda ml: ml.lot_id.id == lot.id):
-                if line.produce_line_ids:
-                    # Do the same process for lot_id contained in produce_line_ids,
-                    # to fetch the end product deliveries
+
+            if moves_by_lot.get(lot.id):
+                producing_move_lines = self.env['stock.move.line'].browse(moves_by_lot[lot.id]['producing_lines'])
+                barren_move_lines = self.env['stock.move.line'].browse(moves_by_lot[lot.id]['barren_lines'])
+
+                if producing_move_lines:
                     lot_path.add(lot.id)
-                    for delivery_ids_set in line.produce_line_ids.lot_id._find_delivery_ids_by_lot(lot_path=lot_path, delivery_by_lot=delivery_by_lot).values():
-                        delivery_ids.update(delivery_ids_set)
-                else:
-                    delivery_ids.add(line.picking_id.id)
+                    next_lots = producing_move_lines.produce_line_ids.lot_id.filtered(lambda l: l.id not in lot_path)
+                    next_lots_ids = set(next_lots.ids)
+                    # If some producing lots are in lot_path, it means that they have been previously processed.
+                    # Their results are therefore already in delivery_by_lot and we add them to delivery_ids directly.
+                    delivery_ids.update(*[set(delivery_by_lot.get(lot_id)) for lot_id in (producing_move_lines.produce_line_ids.lot_id - next_lots).ids])
+
+                    for lot_id, delivery_ids_set in next_lots._find_delivery_ids_by_lot(lot_path=lot_path, delivery_by_lot=delivery_by_lot).items():
+                        if lot_id in next_lots_ids:
+                            delivery_ids.update(delivery_ids_set)
+                delivery_ids.update(barren_move_lines.picking_id.ids)
+
+                # delivery_ids = self._find_delivery_ids(moves_by_lot[lot.id])
```

Comment:
Looks like this line was forgotten to be deleted in the original commit.

Reply:
Hello @ticodoo thanks for the reply :)
I removed the line (also rebased on latest saas-15.1).
Should be ready to `r+` as soon as it's all green :)
Thanks!